src/clj/api/server.clj

```diff
@@ -29,46 +29,52 @@
 ;; NOTE: this is not ideal as every copy of the ECS service will maintain it's own copy
 ;; It would be best to persist it to the RDB
 ;; however as we are currently running only one task per service it will suffice
-(def jailed-ips (atom {}))
-
-(defn update-ip [ip timestamp]
-  (swap! jailed-ips update-in [ip :timestamps] conj timestamp))
-
-(defn is-jailed? [ip now]
-  (let [info       (@jailed-ips ip)
-        timestamps (:timestamps info)]
-    (and info
-         ;; 5 times or more
-         (> (count timestamps) 4)
-         ;; within 10 minutes
-         (< (- now (last timestamps)) (* 10 60 1000))
-         (not (:jail-lift-time info))
-         ;; jailed for 20 minutes
-         (swap! jailed-ips update-in [ip :jail-lift-time] (fn [_] (+ now (* 20 60 1000)))))))
-
-(defn is-temporarily-jailed? [ip now]
-  (let [info           (@jailed-ips ip)
-        jail-lift-time (:jail-lift-time info)]
-    (and info
-         jail-lift-time
-         (> now jail-lift-time)
-         (swap! jailed-ips update-in [ip] dissoc :jail-lift-time))))
+(def ips (atom {}))
+
+;; 1 minute
+;; if within this time period the sam eIP is seen 4 times it will be banned
+(def offending-time-window-length (* 1 60 1000))
+
+;; 2 minutes
+;; ip is banned for this long
+(def jail-sentence (* 2 60 1000))
+
+(defn init-state [ip]
+  {:ip ip :timestamps [] :state :free})
+
+(defn log-ip! [ip timestamp]
+  (swap! ips update-in [ip :timestamps] conj timestamp))
```

Comment:
nice, on thing I find confusing is since timestamps are a list, which will conj in reverse order, then doing (last timestamps) is in fact the first one, which is confusing to read. I guess a (fnil conj []) here instead of the conj, and then you can do (first timestamps)

Reply:
resolving as per discussion in DMs